quartz/depgraph.ts

```diff
@@ -0,0 +1,189 @@
+export default class DepGraph<T> {
+  // node: incoming and outgoing edges
+  _graph = new Map<T, { incoming: Set<T>; outgoing: Set<T> }>()
+
+  constructor() {
+    this._graph = new Map()
+  }
+
+  export(): Object {
+    return {
+      nodes: this.nodes,
+      edges: this.edges,
+    }
+  }
+
+  toString(): string {
+    return JSON.stringify(this.export(), null, 2)
+  }
+
+  // BASIC GRAPH OPERATIONS
+
+  get nodes(): T[] {
+    return Array.from(this._graph.keys())
+  }
+
+  get edges(): [T, T][] {
+    let edges: [T, T][] = []
+    for (const [node, { outgoing }] of this._graph.entries()) {
+      outgoing.forEach((target) => {
+        edges.push([node, target])
+      })
+    }
+    return edges
+  }
+
+  hasNode(node: T): boolean {
+    return this._graph.has(node)
+  }
+
+  addNode(node: T): void {
+    if (!this._graph.has(node)) {
+      this._graph.set(node, { incoming: new Set(), outgoing: new Set() })
+    }
+  }
+
+  removeNode(node: T): void {
+    if (this._graph.has(node)) {
+      this._graph.delete(node)
+    }
+  }
+
+  hasEdge(from: T, to: T): boolean {
+    return this._graph.has(from) && this._graph.get(from)!.outgoing.has(to)
+  }
+
+  addEdge(from: T, to: T): void {
+    this.addNode(from)
```

Comment:
let's make `addNode` return the node so we can use that here and avoid the `!`

Reply:
We'd have to return the entire value of the map, i.e. of `{ incoming: Set<T>; outgoing: Set<T>}` since the function is adding to the incoming and outgoing sets. That would be an unintuitive return value, so leaving as is.